File: sample-bots/csharp/Crazy/Crazy.cs
using System.Drawing;
using Robocode.TankRoyale.BotApi;
using Robocode.TankRoyale.BotApi.Events;

// ------------------------------------------------------------------
// Crazy
// ------------------------------------------------------------------
// A sample bot original made for Robocode by Mathew Nelson.
// Ported to Robocode Tank Royale by Flemming N. Larsen.
//
// This bot moves around in a crazy pattern.
// ------------------------------------------------------------------
public class Crazy : Bot
{
    bool movingForward;

    // The main method starts our bot
    static void Main()
    {
        new Crazy().Start();
    }

    // Constructor, which loads the bot config file
    Crazy() : base(BotInfo.FromFile("Crazy.json")) { }

    // Called when a new round is started -> initialize and do some movement
    public override void Run()
    {
        BodyColor = Color.FromArgb(0x00, 0xC8, 0x00);   // lime
        TurretColor = Color.FromArgb(0x00, 0x96, 0x32); // green
        RadarColor = Color.FromArgb(0x00, 0x64, 0x64);  // dark cyan
        BulletColor = Color.FromArgb(0xFF, 0xFF, 0x64); // yellow
        ScanColor = Color.FromArgb(0xFF, 0xC8, 0xC8);   // light red

        movingForward = true;

        // Loop while as long as the bot is running
        while (IsRunning)
        {
            // Tell the game we will want to move ahead 40000 -- some large number
            SetForward(40000);
            movingForward = true;
            // Tell the game we will want to turn right 90
            SetTurnRight(90);
            // At this point, we have indicated to the game that *when we do something*,
            // we will want to move ahead and turn right. That's what "Set" means.
            // It is important to realize we have not done anything yet!
            // In order to actually move, we'll want to call a method that takes real time, such as
            // WaitFor.
            // WaitFor actually starts the action -- we start moving and turning.
            // It will not return until we have finished turning.
            WaitFor(new TurnCompleteCondition(this));
            // Note: We are still moving ahead now, but the turn is complete.
            // Now we'll turn the other way...
            SetTurnLeft(180);
            // ... and wait for the turn to finish ...
            WaitFor(new TurnCompleteCondition(this));
            // ... then the other way ...
            SetTurnRight(180);
            // ... and wait for that turn to finish.
            WaitFor(new TurnCompleteCondition(this));
            // then back to the top to do it all again.
        }
    }

    // We collided with a wall -> reverse the direction
    public override void OnHitWall(HitWallEvent e)
    {
        // Bounce off!
        ReverseDirection();
    }

    // ReverseDirection: Switch from ahead to back & vice versa
    public void ReverseDirection()
    {
        if (movingForward)
        {
            SetBack(40000);
            movingForward = false;
        }
        else
        {
            SetForward(40000);
            movingForward = true;
        }
    }

    // We scanned another bot -> fire!
    public override void OnScannedBot(ScannedBotEvent e)
    {
        Fire(1);
    }

    // We hit another bot -> back up!
    public override void OnHitBot(HitBotEvent e)
    {
        // If we're moving into the other bot, reverse!
        if (e.IsRammed)
        {
            ReverseDirection();
        }
    }
}

// Condition that is triggered when the turning is complete
public class TurnCompleteCondition : Condition
{
    private readonly Bot bot;

    public TurnCompleteCondition(Bot bot)
    {
        this.bot = bot;
    }

    public override bool Test()
    {
        // turn is complete when the remainder of the turn is zero
        return bot.TurnRemaining == 0;
    }
}

File: sample-bots/python/Crazy/Crazy.py
from robocode_tank_royale.bot_api.bot import Bot
from robocode_tank_royale.bot_api.color import Color
from robocode_tank_royale.bot_api.events import (
    ScannedBotEvent,
    HitWallEvent,
    HitBotEvent,
)


# ------------------------------------------------------------------
# Crazy
# ------------------------------------------------------------------
# A sample bot originally made for Robocode by Mathew Nelson.
#
# This robot moves in a zigzag pattern while firing at enemies.
# ------------------------------------------------------------------
class Crazy(Bot):
    """Moves in a zigzag pattern and fires when scanning other bots."""

    def __init__(self) -> None:
        super().__init__()
        self._moving_forward: bool = False

    def run(self) -> None:
        """Called when a new round is started -> initialize and do some movement."""
        # Set colors (matching the Java sample)
        self.body_color = Color.from_rgb(0x00, 0xC8, 0x00)   # lime
        self.turret_color = Color.from_rgb(0x00, 0x96, 0x32) # green
        self.radar_color = Color.from_rgb(0x00, 0x64, 0x64)  # dark cyan
        self.bullet_color = Color.from_rgb(0xFF, 0xFF, 0x64) # yellow
        self.scan_color = Color.from_rgb(0xFF, 0xC8, 0xC8)   # light red

        # Loop while the bot is running
        while self.running:
            # Tell the game we will want to move ahead 40000 -- some large number
            self.set_forward(40000)
            self._moving_forward = True

            # Tell the game we will want to turn left 90
            self.set_turn_left(90)

            # Start the action and wait until the turn is complete
            self.wait_for(lambda: self.turn_remaining == 0)

            # Note: We are still moving ahead now, but the turn is complete.
            # Now we'll turn the other way...
            self.set_turn_right(180)
            # ... and wait for the turn to finish ...
            self.wait_for(lambda: self.turn_remaining == 0)

            # ... then the other way ...
            self.set_turn_left(180)
            # ... and wait for that turn to finish.
            self.wait_for(lambda: self.turn_remaining == 0)
            # then back to the top to do it all again.

    def on_hit_wall(self, e: HitWallEvent) -> None:
        """We collided with a wall -> reverse the direction."""
        del e
        self._reverse_direction()

    def on_hit_bot(self, e: HitBotEvent) -> None:
        """We hit another bot -> back up if we rammed it."""
        if e.rammed:
            self._reverse_direction()

    def _reverse_direction(self) -> None:
        """Switch from ahead to back & vice versa and commit the change."""
        if self._moving_forward:
            self.set_back(40000)
            self._moving_forward = False
        else:
            self.set_forward(40000)
            self._moving_forward = True
        # Commit the command changes immediately
        self.go()

    def on_scanned_bot(self, e: ScannedBotEvent) -> None:
        """We scanned another bot -> fire!"""
        del e
        self.fire(1)


def main() -> None:
    bot = Crazy()
    bot.start()


if __name__ == "__main__":
    main()
